src/service/ratelimit.go

```diff
@@ -126,11 +134,85 @@ func (this *service) shouldRateLimitWorker(
 			finalCode = descriptorStatus.Code
 		}
 	}
+	if this.responseHeadersEnabled {
+		now := this.clock.UnixSeconds()
+		var limitingDescriptor *pb.RateLimitResponse_DescriptorStatus
+		limitCount := 0
+		for _, descriptor := range responseDescriptorStatuses {
+			if descriptor.CurrentLimit == nil {
+				continue
+			}
+			limitCount++
+			if limitingDescriptor == nil ||
+				descriptor.LimitRemaining < limitingDescriptor.LimitRemaining ||
+				descriptor.LimitRemaining == limitingDescriptor.LimitRemaining &&
+					calculateReset(descriptor, now) > calculateReset(limitingDescriptor, now) {
+				limitingDescriptor = descriptor
+			}
+		}
+		if limitCount == 1 {
+			response.Headers = []*core.HeaderValue{
+				rateLimitLimitHeader(limitingDescriptor),
+				rateLimitRemainingHeader(limitingDescriptor),
+				rateLimitResetHeader(limitingDescriptor, now),
+			}
+		} else if limitCount > 1 {
+			// If there is more than one limit, then picking one of them for the "X-RateLimit-Limit"
```

Comment:
@armsnyder I just ran into this same situation (two or more descriptors that match the request) while building my client library. My test case is contrived and might not represent real world usage, but do we have any sense as to whether most users of this project have multiple descriptors that match on a regular basis?

Like you did here, I have no obvious way to decide which set of limiting conditions to show. Maybe the closest to the limit on a literal or percentage basis - the worst case of them all?

The consuming app doesn't know which rate limiting condition they are hitting when given just the headers and I've never seen a provider that enumerated which rate limit I was hitting.